src/main/scala/mesosphere/marathon/core/launchqueue/impl/AppTaskLauncherActor.scala

```diff
@@ -0,0 +1,346 @@
+  /** Currently known running tasks or tasks that we have requested to be launched. */
+  private[this] var runningTasks: Set[MarathonTask] = _
+  /** Like runningTasks but indexed by the taskId String */
+  private[this] var runningTasksMap: Map[String, MarathonTask] = _
+
+  /** Decorator to use this actor as a [[base.OfferMatcher#TaskLaunchSource]] */
+  private[this] val myselfAsLaunchSource = ActorTaskLaunchSource(self)
+
+  override def preStart(): Unit = {
+    super.preStart()
+
+    log.info("Started appTaskLaunchActor for {} version {} with initial count {}",
+      app.id, app.version, tasksToLaunch)
+
+    taskStatusUpdateSubscription = taskStatusObservable.forAppId(app.id).subscribe { update =>
+      log.debug("update {}", update)
+      self ! update
+    }
+    runningTasks = taskTracker.get(app.id)
+    runningTasksMap = runningTasks.map(task => task.getId -> task).toMap
+
+    rateLimiterActor ! RateLimiterActor.GetDelay(app)
+  }
+
+  override def postStop(): Unit = {
+    taskStatusUpdateSubscription.unsubscribe()
+    OfferMatcherRegistration.unregister()
+
+    super.postStop()
+
+    log.info("Stopped appTaskLaunchActor for {} version {}", app.id, app.version)
+  }
+
+  override def receive: Receive = waitForInitialDelay
+
+  private[this] def waitForInitialDelay: Receive = LoggingReceive.withLabel("waitingForInitialDelay") {
+    case RateLimiterActor.DelayUpdate(delayApp, delayUntil) if delayApp == app =>
+      stash()
+      unstashAll()
+      context.become(active)
+    case message: Any => stash()
+  }
+
+  private[this] def active: Receive = LoggingReceive.withLabel("active") {
+    Seq(
+      receiveStop,
+      receiveDelayUpdate,
+      receiveTaskLaunchNotification,
+      receiveTaskStatusUpdate,
+      receiveGetCurrentCount,
+      receiveAddCount,
+      receiveProcessOffers
+    ).reduce(_.orElse[Any, Unit](_))
+  }
+
+  private[this] def receiveWaitingForInFlight: Receive = LoggingReceive.withLabel("waitingForInFlight") {
+    case notification: TaskLaunchNotification =>
+      receiveTaskLaunchNotification(notification)
+      waitingForInFlight()
+
+    case "waitingForInFlight" => sender() ! "waitingForInFlight" // for testing
+  }
+
+  private[this] def receiveStop: Receive = {
+    case AppTaskLauncherActor.Stop => waitingForInFlight()
+  }
+
+  private[this] def waitingForInFlight(): Unit = {
+    if (inFlightTaskLaunches.isEmpty) {
+      context.stop(self)
+    }
+    else {
+      context.become(receiveWaitingForInFlight)
+    }
+  }
+
+  /**
+    * Receive rate limiter updates.
+    */
+  private[this] def receiveDelayUpdate: Receive = {
+    case RateLimiterActor.DelayUpdate(delayApp, delayUntil) if delayApp == app =>
+      if (backOffUntil.forall(_ < delayUntil)) {
+        import context.dispatcher
+        val now: Timestamp = clock.now()
+        context.system.scheduler.scheduleOnce(now until delayUntil, self, RecheckIfBackOffUntilReached)
```

Comment:
Does'nt that mean, that there is potentially a recheck already scheduled, that should be canceled?